File: sdk-js/src/capability.ts
// Capability engine — per-value provenance, the CaMeL/FIDES core. JS port of
// voan/capability.py. Every value carries an integrity label (was it derived only
// from trusted input, or did untrusted data touch it?) and a confidentiality label
// (which sink classes may receive it); labels propagate through operations. Two
// invariants are enforced per value, deterministically, with no model in the gate:
//   integrity  — an untrusted value may not be a control-sensitive arg of a sink;
//   confidentiality — a value may only reach a sink whose class is in its readers.

export const TRUSTED = "trusted";
export const UNTRUSTED = "untrusted";
export const ANY = "*";

const SENSITIVE_PARAMS = ["recipient", "receiver", "payee", "to", "dest",
  "destination", "address", "account", "iban", "url", "command", "cmd", "path",
  "channel", "user", "query"];

export class Denied extends Error {
  reason: string;
  constructor(reason: string) { super(reason); this.name = "Denied"; this.reason = reason; }
}

function intersect(a: Set<string>, b: Set<string>): Set<string> {
  if (a.has(ANY)) return new Set(b);
  if (b.has(ANY)) return new Set(a);
  return new Set([...a].filter((x) => b.has(x)));
}

export class Capsule {
  value: unknown;
  integrity: string;
  readers: Set<string>;
  source: string | null;

  constructor(value: unknown, integrity: string = TRUSTED,
              readers: Set<string> = new Set([ANY]), source: string | null = null) {
    this.value = value;
    this.integrity = integrity;
    this.readers = readers;
    this.source = source;
  }

  /** A value extracted or computed FROM this one inherits its capability. */
  derive(newValue: unknown, source?: string): Capsule {
    return new Capsule(newValue, this.integrity, new Set(this.readers), source ?? this.source);
  }

  /** Capability of a value computed from several: untrusted if any input is;
   *  readers = intersection (only sinks all inputs permit). */
  static combine(newValue: unknown, capsules: unknown[], source: string | null = null): Capsule {
    const caps = capsules.filter((c): c is Capsule => c instanceof Capsule);
    if (!caps.length) return new Capsule(newValue, TRUSTED, new Set([ANY]), source);
    const integrity = caps.every((c) => c.integrity === TRUSTED) ? TRUSTED : UNTRUSTED;
    let readers = caps[0].readers;
    for (const c of caps.slice(1)) readers = intersect(readers, c.readers);
    return new Capsule(newValue, integrity, readers, source);
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type Fn = (...a: any[]) => any;
interface Step { tool: string; args?: Record<string, unknown>; var?: string }

export class CapabilityEngine {
  sinkClass: Record<string, string>;
  private sensitive: string[];

  constructor(sinkClass: Record<string, string> = {}, sensitive: string[] = SENSITIVE_PARAMS) {
    this.sinkClass = sinkClass;
    this.sensitive = sensitive;
  }

  trusted(value: unknown): Capsule {
    return new Capsule(value, TRUSTED, new Set([ANY]), "user");
  }

  source(tool: string, untrusted = true, readers = new Set([ANY])): Capsule {
    return new Capsule(null, untrusted ? UNTRUSTED : TRUSTED, readers, tool);
  }

  private isSensitive(key: string): boolean {
    return this.sensitive.some((s) => key.toLowerCase().includes(s));
  }

  /** Throw Denied if any Capsule arg violates an invariant, else true. */
  checkCall(tool: string, args: Record<string, unknown>): boolean {
    const cls = this.sinkClass[tool];
    for (const [k, v] of Object.entries(args ?? {})) {
      if (!(v instanceof Capsule)) continue;
      if (v.integrity === UNTRUSTED && this.isSensitive(k)) {
        throw new Denied(`untrusted value steers '${k}' of ${tool} (provenance: ${v.source}) — hijack`);
      }
      if (cls !== undefined && !v.readers.has(ANY) && !v.readers.has(cls)) {
        throw new Denied(`confidential value (readers=${[...v.readers]}) may not reach '${cls}' sink ${tool} — exfiltration`);
      }
    }
    return true;
  }

  private tagOutput(tool: string, args: Record<string, unknown>, result: unknown,
                    untrusted: boolean, readers: Set<string>): Capsule {
    const inCaps = Object.values(args).filter((v) => v instanceof Capsule);
    const combined = Capsule.combine(result, inCaps, tool);
    const integ = (untrusted || combined.integrity === UNTRUSTED) ? UNTRUSTED : TRUSTED;
    const rd = readers.has(ANY) && readers.size === 1 ? combined.readers : intersect(combined.readers, readers);
    return new Capsule(result, integ, rd, tool);
  }

  /** Interpret a capability-tracked PROGRAM (the CaMeL execution model). Each step
   *  {tool, args, var?}; an arg "$name" resolves to a prior step's capsule. */
  run(program: Step[], tools: Record<string, Fn>, sources?: Set<string>,
      confidential: Record<string, Set<string>> = {}, env: Record<string, unknown> = {}): Record<string, unknown> {
    const e = { ...env };
    const srcs = sources ?? new Set(Object.keys(tools));
    for (const step of program) {
      const args = Object.fromEntries(Object.entries(step.args ?? {}).map(
        ([k, v]) => [k, resolve(v, e)]));
      this.checkCall(step.tool, args);
      const raw = Object.fromEntries(Object.entries(args).map(
        ([k, v]) => [k, v instanceof Capsule ? v.value : v]));
      const out = this.tagOutput(step.tool, args, tools[step.tool](raw),
        srcs.has(step.tool), confidential[step.tool] ?? new Set([ANY]));
      if (step.var) e[step.var] = out;
    }
    return e;
  }
}

function resolve(value: unknown, env: Record<string, unknown>): unknown {
  if (typeof value === "string" && value.startsWith("$")) {
    const [base, field] = value.slice(1).split(".");
    const cap = env[base];
    if (cap === undefined) return value;
    if (field && cap instanceof Capsule && cap.value && typeof cap.value === "object") {
      return cap.derive((cap.value as Record<string, unknown>)[field] ?? "");
    }
    return cap;
  }
  return value;
}

File: voan/capability.py
"""Capability engine — per-value provenance, the CaMeL / FIDES core.

The tiers elsewhere in Voan approximate provenance with string matching. This is
the real thing: every value carries a CAPABILITY — an integrity label (was it
derived only from trusted input, or did untrusted data touch it?) and a
confidentiality label (which sink classes may receive it). Capabilities PROPAGATE
through operations: combine a trusted and an untrusted value and the result is
untrusted; combine two values and the readers are the intersection. Two invariants
are then enforced deterministically, per value, and cannot be prompted away:

  INTEGRITY (anti-hijack): an UNTRUSTED value may not be a control-sensitive
    argument of a side effect (a recipient, destination, or command). If the payee
    of a transfer was derived from an email, an injection controlled it — deny.
  CONFIDENTIALITY (anti-exfil): a value may only reach a sink whose class is in its
    readers set. A CONFIDENTIAL value routed to an EXTERNAL sink is denied.

This is the provable property CaMeL gets from a capability interpreter. Agents that
thread Capsules through their tool calls get that guarantee for those values; the
`Firewall(taint=/flow=)` tiers are the automatic, approximate version for agents
that don't. (CaMeL: Debenedetti et al. 2025; FIDES: Microsoft 2025.)
"""
TRUSTED, UNTRUSTED = "trusted", "untrusted"
ANY = "*"   # readers = {ANY} means "may flow anywhere"

# argument names that steer WHERE/HOW a side effect acts — untrusted data here is
# a hijack, so these are the control-sensitive sinks for the integrity invariant.
SENSITIVE_PARAMS = ("recipient", "receiver", "payee", "to", "dest", "destination",
                    "address", "account", "iban", "url", "command", "cmd", "path",
                    "channel", "user", "query")


class Denied(Exception):
    def __init__(self, reason):
        self.reason = reason
        super().__init__(reason)


class Capsule:
    """A value tagged with a capability (integrity + confidentiality readers)."""
    __slots__ = ("value", "integrity", "readers", "source")

    def __init__(self, value, integrity=TRUSTED, readers=frozenset({ANY}), source=None):
        self.value = value
        self.integrity = integrity
        self.readers = frozenset(readers)
        self.source = source

    def derive(self, new_value, source=None):
        """A value extracted or computed FROM this one inherits its capability."""
        return Capsule(new_value, self.integrity, self.readers, source or self.source)

    @staticmethod
    def combine(new_value, *capsules, source=None):
        """Capability of a value computed from several: untrusted if ANY input is,
        readers = intersection (only sinks all inputs permit)."""
        caps = [c for c in capsules if isinstance(c, Capsule)]
        if not caps:
            return Capsule(new_value, TRUSTED, frozenset({ANY}), source)
        integrity = TRUSTED if all(c.integrity == TRUSTED for c in caps) else UNTRUSTED
        readers = caps[0].readers
        for c in caps[1:]:
            readers = _intersect(readers, c.readers)
        return Capsule(new_value, integrity, readers, source)

    def __repr__(self):
        return f"Capsule({self.value!r}, {self.integrity}, readers={set(self.readers)})"


def _intersect(a, b):
    if ANY in a:
        return b
    if ANY in b:
        return a
    return a & b


class CapabilityEngine:
    """Enforces the integrity + confidentiality invariants on tool calls whose
    arguments are Capsules. `sink_class` maps a tool to its sink class (e.g.
    'external'); `readers_for` gives the confidentiality of a source tool's output."""

    def __init__(self, sink_class=None, sensitive_params=SENSITIVE_PARAMS):
        self.sink_class = dict(sink_class or {})
        self.sensitive = tuple(sensitive_params)

    def trusted(self, value):
        return Capsule(value, TRUSTED, frozenset({ANY}), "user")

    def source(self, tool, untrusted=True, readers=frozenset({ANY})):
        """A capsule for data returned by `tool`. Tool outputs are UNTRUSTED by
        default (that is where injections live); pass readers to mark it
        confidential (e.g. readers={'internal'} means 'internal sinks only')."""
        integ = UNTRUSTED if untrusted else TRUSTED
        return Capsule(None, integ, readers, tool)

    def _is_sensitive(self, key):
        return any(s in str(key).lower() for s in self.sensitive)

    def check_call(self, tool, args):
        """Raise Denied if any Capsule argument violates an invariant, else return
        True. Non-Capsule args are treated as trusted literals (the caller vouches)."""
        cls = self.sink_class.get(tool)
        for k, v in (args or {}).items():
            if not isinstance(v, Capsule):
                continue
            if v.integrity == UNTRUSTED and self._is_sensitive(k):
                raise Denied(f"untrusted value steers '{k}' of {tool} "
                             f"(provenance: {v.source}) — hijack")
            if cls is not None and ANY not in v.readers and cls not in v.readers:
                raise Denied(f"confidential value (readers={set(v.readers)}) may not "
                             f"reach '{cls}' sink {tool} — exfiltration")
        return True

    def _tag_output(self, tool, args, result, source, readers):
        """Capability of a tool's OUTPUT: untrusted if this tool is an untrusted
        source OR any input was untrusted (taint propagates through the call);
        readers = intersection of the inputs' readers and this tool's own."""
        in_caps = [v for v in args.values() if isinstance(v, Capsule)]
        combined = Capsule.combine(result, *in_caps, source=tool)
        integ = UNTRUSTED if (source or combined.integrity == UNTRUSTED) else TRUSTED
        rd = combined.readers if readers == frozenset({ANY}) else _intersect(combined.readers, readers)
        return Capsule(result, integ, rd, tool)

    def guard(self, fn, tool, untrusted=True, readers=frozenset({ANY})):
        """Wrap a tool so (1) its Capsule arguments are checked against both
        invariants before it runs, and (2) its RESULT is returned as a Capsule whose
        label propagates this tool's provenance AND its untrusted inputs — so taint
        flows into whatever the agent does next."""
        def wrapped(**kwargs):
            self.check_call(tool, kwargs)
            raw = {k: (v.value if isinstance(v, Capsule) else v) for k, v in kwargs.items()}
            return self._tag_output(tool, kwargs, fn(**raw), untrusted, readers)
        wrapped.__name__ = getattr(fn, "__name__", tool)
        return wrapped

    def guard_tools(self, tools, sources=None, confidential=None):
        """Guard a dict {name: fn}. `sources` names tools whose output is untrusted
        (default: all — tool output is the injection surface). `confidential` maps a
        tool -> its readers set (its output may only reach those sink classes)."""
        sources = set(tools if sources is None else sources)
        confidential = confidential or {}
        return {n: self.guard(f, n, untrusted=(n in sources),
                              readers=confidential.get(n, frozenset({ANY})))
                for n, f in tools.items()}

    def run(self, program, tools, sources=None, confidential=None, env=None):
        """Interpret a capability-tracked PROGRAM — the CaMeL execution model. This
        removes the last gap of the wrapped-tool version: the agent emits a plan of
        steps and Voan runs it, threading capabilities AUTOMATICALLY so no untrusted
        value can escape into a sensitive sink.

        Each step is {tool, args, var?}. An arg value of "$name" resolves to the
        capsule a prior step stored in `name` (so data read from a tool carries its
        untrusted/confidential label into later steps); any other value is a trusted
        literal. `sources` = tools whose own output is untrusted (default: all);
        `confidential` = {tool: readers}. Raises Denied on the first violation.
        Returns the env mapping var -> Capsule."""
        env = dict(env or {})
        srcs = set(tools if sources is None else sources)
        conf = confidential or {}
        for step in program:
            tool = step["tool"]
            args = {k: self._resolve(v, env) for k, v in step.get("args", {}).items()}
            self.check_call(tool, args)
            raw = {k: (v.value if isinstance(v, Capsule) else v) for k, v in args.items()}
            out = self._tag_output(tool, args, tools[tool](**raw),
                                   tool in srcs, conf.get(tool, frozenset({ANY})))
            if step.get("var"):
                env[step["var"]] = out
        return env

    @staticmethod
    def _resolve(value, env):
        if isinstance(value, str) and value.startswith("$"):
            base, _, field = value[1:].partition(".")   # supports $var and $var.field
            cap = env.get(base)
            if cap is None:
                return value
            if field and isinstance(cap, Capsule) and isinstance(cap.value, dict):
                return cap.derive(cap.value.get(field, ""))   # field inherits the taint
            return cap
        return value                               # trusted literal
